File: pc/BmsCompanionBridge/Demo/DemoSource.cs
using System.Reflection;
using BmsCompanion.Bridge.Bms;
using BmsCompanion.Bridge.Server;

namespace BmsCompanion.Bridge.Demo;

/// <summary>
/// Synthetic mission (Hellas, SEAD from Larissa towards Tirana) so the app can be tried and developed without BMS running.
/// Everything is generated; no data is read from the BMS install.
/// </summary>
public sealed class DemoSource
{
    private const double FtPerNm = 6076.12;
    private readonly DateTime _start = DateTime.UtcNow;
    private const double TimeScale = 6; // demo time runs 6x faster than real time

    private static readonly (double x, double y, string desc)[] Route =
    {
        (2293126, 962235, "Takeoff"), (2492326, 741035, "Holding Pt"), (2528258, 695069, "Push"), (2705234, 491385, "Pre IP"),
        (2741700, 452298, "IP"), (2901884, 261470, "Grnd Attack"), (2719964, 456458, "Turn Pt"), (2527725, 693932, "Split"),
        (2414037, 829372, "--"), (2293126, 962235, "Land"), (2129284, 1054691, "Alternate"),
    };

    private static readonly (double x, double y) Bullseye = (2480000, 700000);
    private static readonly (double x, double y) SamSite = (2910660, 258538);
    private static readonly (double x, double y) AwacsOrbit = (2230000, 800000);

    public string BriefingText { get; } = LoadResource("demo_briefing.txt");
    public long BriefingModified { get; } = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();

    private static string LoadResource(string name)
    {
        using var s = Assembly.GetExecutingAssembly().GetManifestResourceStream(name);
        if (s == null) return "";
        using var r = new StreamReader(s);
        return r.ReadToEnd();
    }

    private double SimSeconds => (DateTime.UtcNow - _start).TotalSeconds * TimeScale;

    /// <summary>Position along the route (loops), returns point, heading and the index of the next steerpoint.</summary>
    private static (double x, double y, double hdg, int next) Along(double distFt, int lastLeg = 9)
    {
        double total = 0;
        for (int i = 0; i < lastLeg; i++) total += Dist(Route[i], Route[i + 1]);
        distFt %= total;
        for (int i = 0; i < lastLeg; i++)
        {
            var d = Dist(Route[i], Route[i + 1]);
            if (distFt <= d)
            {
                var f = distFt / d;
                var x = Route[i].x + (Route[i + 1].x - Route[i].x) * f;
                var y = Route[i].y + (Route[i + 1].y - Route[i].y) * f;
                var hdg = (Math.Atan2(Route[i + 1].y - Route[i].y, Route[i + 1].x - Route[i].x) * 180 / Math.PI + 360) % 360;
                return (x, y, hdg, i + 2);
            }
            distFt -= d;
        }
        return (Route[0].x, Route[0].y, 0, 2);
    }

    private static double Dist((double x, double y, string) a, (double x, double y, string) b) => Math.Sqrt((a.x - b.x) * (a.x - b.x) + (a.y - b.y) * (a.y - b.y));

    public LiveDto Live()
    {
        double t = SimSeconds;
        const double gsKts = 480;
        var (x, y, hdg, next) = Along(t * gsKts * FtPerNm / 3600);
        double fuelBurn = (t % 2300) * 2.2;
        var live = new LiveDto
        {
            T = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds(),
            Flying = true,
            Theater = "Hellas",
            Aircraft = "F-16C Block 52+",
            X = x, Y = y, AltFt = 22000 + 400 * Math.Sin(t / 50), HdgTrue = hdg, HdgMag = (hdg - 4 + 360) % 360,
            Kias = 355, Mach = 0.86, GsKts = gsKts, VviFpm = 80 * Math.Cos(t / 50), GLoad = 1.0, Aoa = 3.1, RadarAltFt = 0,
            FuelInternal = Math.Max(1200, 7100 - fuelBurn), FuelExternal = Math.Max(0, 1800 - fuelBurn), FuelFlow = 4800, Bingo = 2400,
            Chaff = 60, Flares = 30, Gear = 0, SpeedBrake = 0,
            BullX = Bullseye.x, BullY = Bullseye.y,
            TimeSec = (int)(3 * 3600 + 43 * 60 + t) % 86400,
            Tacan = "27X", BeaconBrg = 120, BeaconNm = 40, DesiredCourse = 132, NavMode = 2, UhfPreset = 15, UhfFreq = 345950,
            Ded = new List<string>
            {
                $"      STPT  {next,2}  AUTO    ",
                "  LAT  N 40°12.345'     ",
                "  LNG  E 20°51.004'     ",
                "ELEV   22000FT          ",
                $" TOS  {TimeString((int)(3 * 3600 + 58 * 60 + t) % 86400)}          ",
            },
            Voice = new VoiceDto { Flight = "Ouranos5", Seats = "PAXX", Awacs = "Dragnet5", Departure = "Larissa", Arrival = "Larissa", Alternate = "Nea Anchialos" },
        };
        for (int i = 0; i < Route.Length; i++)
            live.NavPoints.Add(new NavPointDto { I = i + 1, Type = "WP", X = Route[i].x, Y = Route[i].y, AltFt = i is 0 or 9 or 10 ? 0 : 22000 });
        live.NavPoints.Add(new NavPointDto { I = 56, Type = "PT", X = SamSite.x, Y = SamSite.y, Name = "SA-5", RangeNm = 60 });
        live.NavPoints.Add(new NavPointDto { I = 57, Type = "PT", X = 2760000, Y = 330000, Name = "SA-6", RangeNm = 14 });
        live.NavPoints.Add(new NavPointDto { I = 26, Type = "MK", X = 2600000, Y = 600000, AltFt = 1200 });

        // RWR: SA-5 site, a MiG-29 CAP and a search radar. Bearings are true, like BMS (the app subtracts heading).
        double Rel(double tx, double ty) => ((Math.Atan2(ty - y, tx - x) * 180 / Math.PI) + 360) % 360;
        live.Rwr.Add(new RwrDto { Sym = 10, Brg = Rel(SamSite.x, SamSite.y), Lethality = 0.7, Lock = (t % 90) > 60 });
        live.Rwr.Add(new RwrDto { Sym = 2, Brg = Rel(2850000, 420000), Lethality = 0.5, New = (t % 40) < 4 });
        live.Rwr.Add(new RwrDto { Sym = 17, Brg = Rel(2980000, 300000), Lethality = 0.2 });
        return live;
    }

    public ContactsDto Contacts()
    {
        double t = SimSeconds;
        var own = Live();
        var dto = new ContactsDto { T = own.T, Connected = true, State = "demo" };
        void Add(string id, string kind, double x, double y, double alt, double hdg, double gs, string name, string? group, string coal, bool friendly, bool isOwn = false) =>
            dto.Contacts.Add(new ContactDto
            {
                Id = id, Kind = kind, X = x, Y = y, AltFt = alt, Hdg = Math.Round(hdg, 1), GsKts = gs, Name = name, Group = group,
                Coalition = coal, Color = friendly ? "Blue" : "Red", Friendly = friendly, Own = isOwn,
            });

        Add("1", "air", own.X, own.Y, own.AltFt, own.HdgTrue, own.GsKts, "F-16CM-52", "Ouranos5", "Greece", true, true);
        var wing = Along(Math.Max(0, t * 480 * FtPerNm / 3600 - 1.5 * FtPerNm));
        Add("2", "air", wing.x + 2500, wing.y - 2500, 21500, wing.hdg, 480, "F-16CM-52", "Ouranos5", "Greece", true);
        var rider = Along(Math.Max(0, t * 470 * FtPerNm / 3600 - 9 * FtPerNm));
        Add("3", "air", rider.x - 6000, rider.y + 4000, 20000, rider.hdg, 470, "F-4E", "Rider5", "Greece", true);
        Add("4", "air", rider.x - 8500, rider.y + 6500, 20500, rider.hdg, 470, "F-4E", "Rider5", "Greece", true);
        double a = t / 400;
        Add("5", "air", AwacsOrbit.x + Math.Cos(a) * 15 * FtPerNm, AwacsOrbit.y + Math.Sin(a) * 25 * FtPerNm, 30000, (a * 180 / Math.PI + 90) % 360, 320, "EMB-145H", "Dragnet5", "Greece", true);
        double m = t / 260;
        Add("10", "air", 2850000 + Math.Cos(m) * 12 * FtPerNm, 420000 + Math.Sin(m) * 20 * FtPerNm, 26000, (m * 180 / Math.PI + 90) % 360, 450, "MiG-29A", "Falcon2", "Albania", false);
        Add("11", "air", 2850000 + Math.Cos(m) * 12 * FtPerNm - 8000, 420000 + Math.Sin(m) * 20 * FtPerNm + 6000, 25000, (m * 180 / Math.PI + 90) % 360, 450, "MiG-29A", "Falcon2", "Albania", false);
        Add("12", "heli", 2560000, 640000 + (t % 600) * 200, 800, 90, 110, "UH-60", "Pedro1", "Greece", true);
        Add("20", "ship", 2400000, 330000, 0, 150, 12, "Elli class frigate", null, "Greece", true);
        // hostile ejected crew close to ownship: the Picture list must still show the MiGs first
        Add("40", "crew", own.X + 4 * FtPerNm, own.Y - 3 * FtPerNm, 0, 0, 0, "Ejected Crew", null, "Albania", false);
        if ((t % 120) < 25)
        {
            double f = (t % 120) / 25;
            Add("30", "missile", SamSite.x + (own.X - SamSite.x) * f * 0.6, SamSite.y + (own.Y - SamSite.y) * f * 0.6, 5000 + 30000 * f, 0, 1800, "SA-5 Gammon", null, "Albania", false);
        }
        Add("90", "bullseye", Bullseye.x, Bullseye.y, 0, 0, 0, "Bullseye", null, "Greece", true);
        return dto;
    }

    public Dtc Dtc()
    {
        var d = new Dtc { Modified = BriefingModified };
        for (int i = 0; i < Route.Length; i++)
            d.Steerpoints.Add(new DtcPoint { N = i + 1, X = Route[i].x, Y = Route[i].y, AltFt = i is 0 or 9 or 10 ? 0 : 22000, Action = i == 5 ? -1 : 1, Name = i == 5 ? "SA-5 launchers, 4 nm S of Tirana" : null });
        d.Ppts.Add(new DtcPpt { N = 56, X = SamSite.x, Y = SamSite.y, RangeNm = 60, Name = "SA-5" });
        d.Ppts.Add(new DtcPpt { N = 57, X = 2760000, Y = 330000, RangeNm = 14, Name = "SA-6" });
        string[] uhf = { "382.500|Base ops", "264.500|DEP Ground", "362.400|DEP Tower", "265.150|DEP Approach", "342.275|AWACS Check-in", "399.125|Tactical", "265.150|ARR Approach", "362.400|ARR Tower", "264.500|ARR Ground", "342.600|ALT Approach", "364.400|ALT Tower", "340.300|ALT Ground", "348.350|", "339.750|Advisory", "345.950|Intra Flight 1", "251.000|Intra Flight 2" };
        for (int i = 0; i < uhf.Length; i++)
        {
            var p = uhf[i].Split('|');
            d.Uhf.Add(new Preset { Ch = i + 1, Freq = p[0], Comment = p[1].Length > 0 ? p[1] : null });
        }
        d.Vhf.Add(new Preset { Ch = 2, Freq = "123.250", Comment = "DEP ATIS" });
        d.Vhf.Add(new Preset { Ch = 3, Freq = "120.550", Comment = "DEP Tower" });
        d.Vhf.Add(new Preset { Ch = 14, Freq = "119.500", Comment = "UNICOM" });
        d.Vhf.Add(new Preset { Ch = 15, Freq = "52.450", Comment = "Flight-1" });
        d.Iff["Mode1 Code"] = "60"; d.Iff["Mode2 Code"] = "1404"; d.Iff["Mode3A Code"] = "7504"; d.Iff["Mode4 Key"] = "1";
        return d;
    }

    private static string TimeString(int s) => $"{s / 3600:00}:{s / 60 % 60:00}:{s % 60:00}";
}

File: desktop/src/main/kotlin/com/bmscompanion/desktop/bridge/DemoSource.kt
package com.bmscompanion.desktop.bridge

import com.bmscompanion.app.data.mission.Contact
import com.bmscompanion.app.data.mission.Contacts
import com.bmscompanion.app.data.mission.Dtc
import com.bmscompanion.app.data.mission.DtcPoint
import com.bmscompanion.app.data.mission.DtcPpt
import com.bmscompanion.app.data.mission.Live
import com.bmscompanion.app.data.mission.NavPoint
import com.bmscompanion.app.data.mission.Preset
import com.bmscompanion.app.data.mission.RwrContact
import com.bmscompanion.app.data.mission.Voice
import kotlin.math.PI
import kotlin.math.atan2
import kotlin.math.cos
import kotlin.math.max
import kotlin.math.pow
import kotlin.math.round
import kotlin.math.sin
import kotlin.math.sqrt

/**
 * Synthetic mission (Hellas, SEAD from Larissa towards Tirana) so everything can be tried without BMS running.
 * Everything is generated; nothing is read from the BMS install.
 */
class DemoSource {
    private val start = System.currentTimeMillis()

    val briefingText: String = DemoSource::class.java.getResourceAsStream("/bridge/demo_briefing.txt")?.use { it.readBytes().toString(Charsets.UTF_8) }.orEmpty()
    val briefingModified: Long = System.currentTimeMillis()

    /** Demo time runs 6x faster than real time. */
    private val simSeconds get() = (System.currentTimeMillis() - start) / 1000.0 * 6

    private data class Wp(val x: Double, val y: Double, val desc: String)

    private val route = listOf(
        Wp(2293126.0, 962235.0, "Takeoff"), Wp(2492326.0, 741035.0, "Holding Pt"), Wp(2528258.0, 695069.0, "Push"), Wp(2705234.0, 491385.0, "Pre IP"),
        Wp(2741700.0, 452298.0, "IP"), Wp(2901884.0, 261470.0, "Grnd Attack"), Wp(2719964.0, 456458.0, "Turn Pt"), Wp(2527725.0, 693932.0, "Split"),
        Wp(2414037.0, 829372.0, "--"), Wp(2293126.0, 962235.0, "Land"), Wp(2129284.0, 1054691.0, "Alternate"),
    )
    private val bullseye = 2480000.0 to 700000.0
    private val samSite = 2910660.0 to 258538.0
    private val awacsOrbit = 2230000.0 to 800000.0

    private data class Along(val x: Double, val y: Double, val hdg: Double, val next: Int)

    /** Position along the route (loops): point, heading and the index of the next steerpoint. */
    private fun along(distance: Double, lastLeg: Int = 9): Along {
        var total = 0.0
        for (i in 0 until lastLeg) total += dist(route[i], route[i + 1])
        var d = distance % total
        for (i in 0 until lastLeg) {
            val leg = dist(route[i], route[i + 1])
            if (d <= leg) {
                val f = d / leg
                val a = route[i]
                val b = route[i + 1]
                return Along(a.x + (b.x - a.x) * f, a.y + (b.y - a.y) * f, (atan2(b.y - a.y, b.x - a.x) * 180 / PI + 360) % 360, i + 2)
            }
            d -= leg
        }
        return Along(route[0].x, route[0].y, 0.0, 2)
    }

    private fun dist(a: Wp, b: Wp) = sqrt((a.x - b.x) * (a.x - b.x) + (a.y - b.y) * (a.y - b.y))

    fun live(): Live {
        val t = simSeconds
        val gsKts = 480.0
        val (x, y, hdg, next) = along(t * gsKts * FT_PER_NM / 3600)
        val fuelBurn = (t % 2300) * 2.2
        fun rel(tx: Double, ty: Double) = ((atan2(ty - y, tx - x) * 180 / PI) + 360) % 360
        val navPoints = route.mapIndexed { i, w -> NavPoint(i + 1, "WP", w.x, w.y, if (i == 0 || i == 9 || i == 10) 0.0 else 22000.0) } + listOf(
            NavPoint(56, "PT", samSite.first, samSite.second, name = "SA-5", rangeNm = 60.0),
            NavPoint(57, "PT", 2760000.0, 330000.0, name = "SA-6", rangeNm = 14.0),
            NavPoint(26, "MK", 2600000.0, 600000.0, 1200.0),
        )
        return Live(
            t = System.currentTimeMillis(), flying = true, theater = "Hellas", aircraft = "F-16C Block 52+",
            x = x, y = y, altFt = 22000 + 400 * sin(t / 50), hdgTrue = hdg, hdgMag = (hdg - 4 + 360) % 360,
            kias = 355.0, mach = 0.86, gsKts = gsKts, vviFpm = 80 * cos(t / 50), gLoad = 1.0, aoa = 3.1, radarAltFt = 0.0,
            fuelInternal = max(1200.0, 7100 - fuelBurn), fuelExternal = max(0.0, 1800 - fuelBurn), fuelFlow = 4800.0, bingo = 2400.0,
            chaff = 60, flares = 30, gear = 0.0, speedBrake = 0.0,
            bullX = bullseye.first, bullY = bullseye.second,
            timeSec = ((3 * 3600 + 43 * 60 + t).toInt()) % 86400,
            tacan = "16Y A/A", tacanUfc = "27X", tacanAux = "16Y A/A", beaconBrg = 120.0, beaconNm = 40.0, desiredCourse = 132.0, navMode = 2, uhfPreset = 15, uhfFreq = 345950,
            ded = listOf(
                "      STPT  ${next.toString().padStart(2)}  AUTO    ",
                "  LAT  N 40°12.345'     ",
                "  LNG  E 20°51.004'     ",
                "ELEV   22000FT          ",
                " TOS  ${timeString(((3 * 3600 + 58 * 60 + t).toInt()) % 86400)}          ",
            ),
            voice = Voice(flight = "Ouranos5", seats = "PAXX", awacs = "Dragnet5", departure = "Larissa", arrival = "Larissa", alternate = "Nea Anchialos"),
            navPoints = navPoints,
            // SA-5 site, a MiG-29 CAP and a search radar. Bearings are true, like BMS (the app subtracts heading).
            rwr = listOf(
                RwrContact(sym = 10, brg = rel(samSite.first, samSite.second), lethality = 0.7, lock = (t % 90) > 60),
                RwrContact(sym = 2, brg = rel(2850000.0, 420000.0), lethality = 0.5, new = (t % 40) < 4),
                RwrContact(sym = 17, brg = rel(2980000.0, 300000.0), lethality = 0.2),
            ),
        )
    }

    fun contacts(): Contacts {
        val t = simSeconds
        val own = live()
        val list = ArrayList<Contact>()
        fun add(id: String, kind: String, x: Double, y: Double, alt: Double, hdg: Double, gs: Double, name: String, group: String?, coalition: String, friendly: Boolean, isOwn: Boolean = false): Int {
            var c = Contact(
                id = id, kind = kind, x = x, y = y, altFt = alt, hdg = round(hdg * 10) / 10, gsKts = gs, name = name, group = group,
                coalition = coalition, color = if (friendly) "Blue" else "Red", friendly = friendly, own = isOwn,
            )
            if (kind == "air" || kind == "heli") c = c.copy(
                ias = round(gs * (if (alt > 15000) 0.72 else 0.92)), // rough IAS from ground speed at altitude
                mach = round(gs / 600.0 * 100) / 100,
                fuelLb = round(max(900.0, 7400 - (t % 3600) * 1.4 - id.length * 300)),
            )
            list += c
            return list.size - 1
        }

        add("1", "air", own.x, own.y, own.altFt, own.hdgTrue, own.gsKts, "F-16CM-52", "Ouranos5", "Greece", true, true)
        val wing = along(max(0.0, t * 480 * FT_PER_NM / 3600 - 1.5 * FT_PER_NM))
        add("2", "air", wing.x + 2500, wing.y - 2500, 21500.0, wing.hdg, 480.0, "F-16CM-52", "Ouranos5", "Greece", true)
        val rider = along(max(0.0, t * 470 * FT_PER_NM / 3600 - 9 * FT_PER_NM))
        add("3", "air", rider.x - 6000, rider.y + 4000, 20000.0, rider.hdg, 470.0, "F-4E", "Rider5", "Greece", true)
        add("4", "air", rider.x - 8500, rider.y + 6500, 20500.0, rider.hdg, 470.0, "F-4E", "Rider5", "Greece", true)
        val a = t / 400
        add("5", "air", awacsOrbit.first + cos(a) * 15 * FT_PER_NM, awacsOrbit.second + sin(a) * 25 * FT_PER_NM, 30000.0, (a * 180 / PI + 90) % 360, 320.0, "EMB-145H", "Dragnet5", "Greece", true)
        val k = t / 520 // tanker racetrack north-east of Larissa
        add("6", "air", 2400000 + cos(k) * 8 * FT_PER_NM, 1080000 + sin(k) * 18 * FT_PER_NM, 24000.0, (k * 180 / PI + 90) % 360, 310.0, "KC-135R", "Texaco1", "Greece", true)
        val m = t / 260
        add("10", "air", 2850000 + cos(m) * 12 * FT_PER_NM, 420000 + sin(m) * 20 * FT_PER_NM, 26000.0, (m * 180 / PI + 90) % 360, 450.0, "MiG-29A", "Falcon2", "Albania", false)
        add("11", "air", 2850000 + cos(m) * 12 * FT_PER_NM - 8000, 420000 + sin(m) * 20 * FT_PER_NM + 6000, 25000.0, (m * 180 / PI + 90) % 360, 450.0, "MiG-29A", "Falcon2", "Albania", false)
        add("12", "heli", 2560000.0, 640000 + (t % 600) * 200, 800.0, 90.0, 110.0, "UH-60", "Pedro1", "Greece", true)
        // AWACS demo: a friendly CAP (Colt1) orbiting west of the bullseye, and a hostile pair (Uzi2) running in from the north-west
        val cap = t / 300
        val coltX = 2600000 + cos(cap) * 10 * FT_PER_NM
        val coltY = 560000 + sin(cap) * 16 * FT_PER_NM
        val coltHdg = (cap * 180 / PI + 90) % 360
        add("50", "air", coltX, coltY, 24000.0, coltHdg, 430.0, "F-16CM-50", "Colt1", "Greece", true)
        add("51", "air", coltX - 9000, coltY + 7000, 23000.0, coltHdg, 430.0, "F-16CM-50", "Colt1", "Greece", true)
        val run = (t % 900) / 900 // restarts every 900 s of demo time
        val uziX = 2950000 - run * 60 * FT_PER_NM * 5.2
        val uziY = 330000 + run * 60 * FT_PER_NM * 3.4
        val uziHdg = atan2(3.4, -5.2) * 180 / PI + 360
        val uzi1 = add("60", "air", uziX, uziY, 31000.0, uziHdg % 360, 540.0, "Su-27", "Uzi2", "Albania", false)
        add("61", "air", uziX + 12000, uziY - 9000, 29000.0, uziHdg % 360, 540.0, "Su-27", "Uzi2", "Albania", false)
        if (sqrt((uziX - coltX).pow(2) + (uziY - coltY).pow(2)) < 45 * FT_PER_NM) list[uzi1] = list[uzi1].copy(locked = "50")
        add("20", "ship", 2400000.0, 330000.0, 0.0, 150.0, 12.0, "Elli class frigate", null, "Greece", true)
        // hostile ejected crew close to ownship: the Picture list must still show the MiGs first
        add("40", "crew", own.x + 4 * FT_PER_NM, own.y - 3 * FT_PER_NM, 0.0, 0.0, 0.0, "Ejected Crew", null, "Albania", false)
        if ((t % 120) < 25) {
            val f = (t % 120) / 25
            add("30", "missile", samSite.first + (own.x - samSite.first) * f * 0.6, samSite.second + (own.y - samSite.second) * f * 0.6, 5000 + 30000 * f, 0.0, 1800.0, "SA-5 Gammon", null, "Albania", false)
        }
        add("90", "bullseye", bullseye.first, bullseye.second, 0.0, 0.0, 0.0, "Bullseye", null, "Greece", true)
        return Contacts(t = own.t, connected = true, state = "demo", contacts = list)
    }

    fun dtc(): Dtc {
        val uhf = listOf(
            "382.500|Base ops", "264.500|DEP Ground", "362.400|DEP Tower", "265.150|DEP Approach", "342.275|AWACS Check-in", "399.125|Tactical",
            "265.150|ARR Approach", "362.400|ARR Tower", "264.500|ARR Ground", "342.600|ALT Approach", "364.400|ALT Tower", "340.300|ALT Ground",
            "348.350|", "339.750|Advisory", "345.950|Intra Flight 1", "251.000|Intra Flight 2",
        ).mapIndexed { i, s -> s.split('|').let { Preset(i + 1, it[0], it[1].ifEmpty { null }) } }
        return Dtc(
            modified = briefingModified,
            steerpoints = route.mapIndexed { i, w ->
                DtcPoint(n = i + 1, x = w.x, y = w.y, altFt = if (i == 0 || i == 9 || i == 10) 0.0 else 22000.0, action = if (i == 5) -1 else 1, isTarget = i == 5,
                    name = if (i == 5) "SA-5 launchers, 4 nm S of Tirana" else null)
            },
            ppts = listOf(DtcPpt(56, samSite.first, samSite.second, 0.0, 60.0, "SA-5"), DtcPpt(57, 2760000.0, 330000.0, 0.0, 14.0, "SA-6")),
            uhf = uhf,
            vhf = listOf(Preset(2, "123.250", "DEP ATIS"), Preset(3, "120.550", "DEP Tower"), Preset(14, "119.500", "UNICOM"), Preset(15, "52.450", "Flight-1")),
            iff = linkedMapOf("Mode1 Code" to "60", "Mode2 Code" to "1404", "Mode3A Code" to "7504", "Mode4 Key" to "1"),
        )
    }

    private fun timeString(s: Int) = "%02d:%02d:%02d".format(s / 3600, s / 60 % 60, s % 60)

    private companion object {
        const val FT_PER_NM = 6076.12
    }
}
